lib/Transforms/InsertGpuAllocs.cpp

```diff
@@ -0,0 +1,363 @@
+                               mlir::OperationPass<mlir::func::FuncOp>> {
+
+  InsertGPUAllocs() = default;
+  InsertGPUAllocs(bool gpuDealloc) { useGpuDealloc = gpuDealloc; }
+  InsertGPUAllocs(const InsertGPUAllocs &pass) : PassWrapper(pass) {}
+
+  virtual void
+  getDependentDialects(mlir::DialectRegistry &registry) const override {
+    registry.insert<mlir::memref::MemRefDialect>();
+    registry.insert<mlir::gpu::GPUDialect>();
+    registry.insert<mlir::scf::SCFDialect>();
+    registry.insert<mlir::arith::ArithmeticDialect>();
+  }
+
+  void runOnOperation() override {
+    auto func = getOperation();
+    auto &funcBody = func.getBody();
+    if (funcBody.empty()) {
+      return;
+    } else if (!llvm::hasSingleElement(funcBody)) {
+      func.emitError("Function must have exactly one block");
+      signalPassFailure();
+      return;
+    }
+
+    struct AccessType {
+      bool hostRead = false;
+      bool hostWrite = false;
+      bool deviceRead = false;
+      bool deviceWrite = false;
+    };
+
+    llvm::SmallMapVector<mlir::Operation *, AccessType, 8> gpuBufferAllocs;
+    llvm::SmallMapVector<unsigned, AccessType, 8> gpuBufferParams;
+    llvm::SmallMapVector<mlir::Operation *, AccessType, 8>
+        gpuGetMemrefGlobalParams;
+    auto &aliases = getAnalysis<mlir::BufferViewFlowAnalysis>();
+
+    auto getMemref = [](mlir::Operation *op)
+        -> llvm::Optional<mlir::SmallVector<mlir::Value, 4>> {
+      if (auto load = mlir::dyn_cast<mlir::memref::LoadOp>(op)) {
+        return {{load.memref()}};
+      } else if (auto store = mlir::dyn_cast<mlir::memref::StoreOp>(op)) {
+        return {{store.memref()}};
+      } else if (auto call = mlir::dyn_cast<mlir::func::CallOp>(op)) {
+        mlir::SmallVector<mlir::Value, 4> ret;
+        for (auto arg : call.operands()) {
+          if (arg.getType().isa<mlir::MemRefType>())
+            ret.emplace_back(arg);
+        }
+        return std::move(ret);
+      } else {
+        op->emitError("Uhhandled mem op in gpu region");
+        return llvm::None;
+      }
+    };
+
+    auto scfDialect = getContext().getOrLoadDialect<mlir::scf::SCFDialect>();
+
+    auto hasMemAccess = [](mlir::Operation *op) -> bool {
+      if (auto memInterface =
+              mlir::dyn_cast<mlir::MemoryEffectOpInterface>(op)) {
+        if (memInterface.hasEffect<mlir::MemoryEffects::Read>() ||
+            memInterface.hasEffect<mlir::MemoryEffects::Write>())
+          return true;
+      }
+      if (auto call = mlir::dyn_cast<mlir::func::CallOp>(op)) {
+        for (auto arg : call.operands()) {
+          if (arg.getType().isa<mlir::MemRefType>())
+            return true;
+        }
+      }
+      return false;
+    };
+
+    // Traverse through all the memory access ops under GPU launch Op
+    // and add device memory allocation appropriately.
+    if (func.walk([&](mlir::Operation *op) {
+              if (!op->getParentOfType<mlir::gpu::LaunchOp>())
```

Comment:
how would it be any different after the kernels are outlined? currently this pass runs before the kernels get outlined. The drawback I see is making sure the gpu level tests have to have the gpu.launch blocks ... format...

Reply:
>>how would it be any different after the kernels are outlined?
I think that may involve cross device func and host code analysis, since the memref used inside device kernel needs to be gpu.alloc in the host code.  I would expect the code needs to handle launch_func.  I don't know how important it is so far for our use cases, but I think we need to tell user upfront that we are not handling gpu.launch_func. 

>> currently this pass runs before the kernels get outlined.
Maybe you can put this in the pass introduction.  So user knows the limitation.

Reply:
This pass was specifically written to be run before kernel outlining, it is much harder to work on outlined kernel and generally not worth it.